cli: Deprecate --wait=live in `node decommission`

neeral opened the pull request:
Refer to issue #26880.

When you try to decommission a node that is down, today one has to use
`decommission --wait=live`, which does not verify whether the down node
is part of any ranges, and manually verify that the cluster has
up-replicated all ranges elsewhere. This is far from ideal, in particular
since there is no automated way to reliably check this.

`--wait=live` is deprecated and its behaviour replaced by that of
`--wait=all`. Instead of relying on metrics for replica counts, which
may be stale, look at authoritive source - range metadata.

Release note (cli change):
Deprecate `--wait=live` parameter for `node decommission`. `--wait=all`
is the default behaviour. This ensures CockroachDB checks ranges are on
the node to be decommissioned are not under-replicated before the node
is decommissioned.

neeral committed:
cli: Deprecate --wait=live in `node decommission`

Refer to issue #26880.

When you try to decommission a node that is down, today one has to use
`decommission --wait=live`, which does not verify whether the down node
is part of any ranges, and manually verify that the cluster has
up-replicated all ranges elsewhere. This is far from ideal, in particular
since there is no automated way to reliably check this.

`--wait=live` is deprecated and its behaviour replaced by that of
`--wait=all`. Instead of relying on metrics for replica counts, which
may be stale, look at authoritive source - range metadata.

Release note (cli change):
Deprecate `--wait=live` parameter for `node decommission`. `--wait=all`
is the default behaviour. This ensures CockroachDB checks ranges on the
node to be decommissioned are not under-replicated before the node is
decommissioned.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/27027)
<!-- Reviewable:end -->

neeral commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/server/admin.go, line 1300 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG1Lqd_D_owXHdVJq-9:-LG1LqdaCf2lzGK4p3xW:bopkjj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a20af9a3e557959da52b887f08c705fa70096f35/pkg/server/admin.go#L1300)):*
> ```Go
> 	// a lookup table to check whether we care about a given node.
> 	replicaCounts := make(map[roachpb.NodeID]int64)
> 	// TODO (neeral) make this paginated
> ```

I'm working on this - I wanted to receive some early feedback on the approach and reasoning (see comment on the issue).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
This looks good! Just needs some cleanup. I think the existing test coverage we have for this is sufficient except that I'd like to see a test of the iteration logic. I think that will come out cleanest if you make it a method on `*Txn`, something like

```go
func (t *Txn) Iterate(ctx context.Context, from, to interface{}, pageSize int, f func(rows []KeyValue)) {
  // Basically the meat of your code
}
```

which you can conveniently test from `client/db_test.go` (via a `TestServer` which has a `SplitRange` method).

---

Reviewed 5 of 5 files at r2.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/acceptance/decommission_test.go, line 378 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6SAVa4hsXKWMFcZf4:-LG6SAVa4hsXKWMFcZf5:bt42sal) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/acceptance/decommission_test.go#L378)):*
> ```Go
> 			{"--wait=live is deprecated and is treated as --wait=all"},
> 			decommissionHeader,
> 			{strconv.Itoa(int(target)), `true|false`, `\d+`, `true`, `true|false`},
> ```

This should be `0` now instead of `\d+`, and probably similar changes are adequate in other places in this test.

---

*[pkg/cli/node.go, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6SL6-EDh0D-DqEgTs:-LG6SL6-EDh0D-DqEgTt:b-rlzfqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/cli/node.go#L320)):*
> ```Go
> ) error {
> 	if wait == nodeDecommissionWaitLive {
> 		fmt.Fprintln(os.Stdout, "\n--wait=live is deprecated and is treated as --wait=all")
> ```

@knz is this the right place to print this? Or should we even print? I'd like to be able to remove that value at some point, though perhaps that's not worth it, and we shouldn't even bother operators?

---

*[pkg/cli/node.go, line 361 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6SX9yDVLnIjeFVnbo:-LG6SX9yDVLnIjeFVnbp:b-pxqhq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/cli/node.go#L361)):*
> ```Go
> 		}
> 		if replicaCount == 0 && allDecommissioning {
> 			fmt.Fprintln(os.Stdout, "\nAll target nodes report that they hold no more data. "+
> ```

Perhaps change this to "No more data reported on target nodes. Please verify...".

---

*[pkg/server/admin.go, line 1306 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6SuXj7DcLJIETIpqh:-LG6SuXk3OgTEYHg4qjX:b-84dz33) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/server/admin.go#L1306)):*
> ```Go
> 		begin := keys.Meta2Prefix
> 		for {
> 			rows, err := s.server.db.Scan(ctx, begin, keys.MetaMax, pageSize)
> ```

You need to use `txn` here or you're not seeing a consistent view. Which I guess is fine.. but then you don't need the txn at all. I would use the txn because we try to allow less range-spanning stuff outside of one now.

---

*[pkg/server/admin.go, line 1325 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6TgbM0sWBLmcSKmCA:-LG6TgbM0sWBLmcSKmCB:b40ulix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/server/admin.go#L1325)):*
> ```Go
> 			}
> 			begin = rows[len(rows)-1].Key.Next()
> 			time.Sleep(time.Second)
> ```

Come to think of it, you might as well remove this Sleep. All it'll do is make unit testing slower.

---

*[pkg/server/serverpb/admin.proto, line 379 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6Tn9pC-OmcvS0EuNF:-LG6Tn9pC-OmcvS0EuNG:bbg99pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/server/serverpb/admin.proto#L379)):*
> ```protobuf
>                         (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/roachpb.NodeID"];
>     bool is_live = 2;
>     // The numiber of replicas on a node, computed by scanning meta2 ranges.
> ```

Typo, and nit: "the node".

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-:-LG6TsVFDZLUi_FiGlMX:b-j32pzk)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/cli/node.go, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6SL6-EDh0D-DqEgTs:-LG7NYuN6Rltt5HlPWnq:b-c9l1ei) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/cli/node.go#L320)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

@knz is this the right place to print this? Or should we even print? I'd like to be able to remove that value at some point, though perhaps that's not worth it, and we shouldn't even bother operators?
</blockquote></details>

It should go to stderr, not stdout. I think it makes sense to have a deprecation message here for one release cycle and then remove it in the next. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027)*
<!-- Sent from Reviewable.io -->

neeral commented:
I've added Txn:Iterate() and added a db_test.go:TestDB_TxnIterate(). I didn't understand your comment about `TestServer` and `SplitRange`. I followed the same type of behaviour as the other tests in the file.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/acceptance/decommission_test.go, line 378 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6SAVa4hsXKWMFcZf4:-LGDKemh1MtWyF1c7rwY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/acceptance/decommission_test.go#L378)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This should be `0` now instead of `\d+`, and probably similar changes are adequate in other places in this test.
</blockquote></details>

Done.

---

*[pkg/cli/node.go, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6SL6-EDh0D-DqEgTs:-LGDFb-q9OKJaN9wVRmK:b-jrwe4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/cli/node.go#L320)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It should go to stderr, not stdout. I think it makes sense to have a deprecation message here for one release cycle and then remove it in the next. 
</blockquote></details>

Changed to using Stderr.

---

*[pkg/cli/node.go, line 361 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6SX9yDVLnIjeFVnbo:-LG6pNNP43XND4XWH9EP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/cli/node.go#L361)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Perhaps change this to "No more data reported on target nodes. Please verify...".
</blockquote></details>

Done.

---

*[pkg/server/admin.go, line 1306 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6SuXj7DcLJIETIpqh:-LG6rBcaEpLhDSrOLUhR:b-ce1np7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/server/admin.go#L1306)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

You need to use `txn` here or you're not seeing a consistent view. Which I guess is fine.. but then you don't need the txn at all. I would use the txn because we try to allow less range-spanning stuff outside of one now.
</blockquote></details>

Oops, i meant to use a txn here. I forgot to replace db with txn after a small refactor. Thanks for catching it. I agree, should use a txn for a consistent view.

---

*[pkg/server/admin.go, line 1325 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6TgbM0sWBLmcSKmCA:-LG6ruIA4VrHKSllsoz7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/server/admin.go#L1325)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Come to think of it, you might as well remove this Sleep. All it'll do is make unit testing slower.
</blockquote></details>

Done.

---

*[pkg/server/serverpb/admin.proto, line 379 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6Tn9pC-OmcvS0EuNF:-LG6s15x5PwpytDwEE7P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/server/serverpb/admin.proto#L379)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Typo, and nit: "the node".
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-:-LGDKrgl4YHcr0z6NrWJ:b-gydhyj)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/cli/node.go, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6SL6-EDh0D-DqEgTs:-LGDMzrg-pEDimzfZiRq:bioer69) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/cli/node.go#L320)):*
<details><summary><i>Previously, neeral (Neeral Dodhia) wrote…</i></summary><blockquote>

Changed to using Stderr.
</blockquote></details>

use `stderr` defined in this package (not `os.Stderr`), this is necessary to test the output properly.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
My comment wasn't useful. I somehow imagined you'd want to test with splits, but that doesn't apply here.

<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 7 of 7 files at r3.
Review status: :shipit: complete! 1 of 0 LGTMs obtained

---

*[pkg/internal/client/db_test.go, line 280 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LGFBKV3AnRbrThsvclb:-LGFBKV3AnRbrThsvclc:b-nptnte) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67d7b4c612e2e037ee0fe29a2ccff709738eb22/pkg/internal/client/db_test.go#L280)):*
> ```Go
> 	p := 0
> 	if err := db.Txn(context.TODO(), func(ctx context.Context, txn *client.Txn) error {
> 		return txn.Iterate(context.TODO(), "a", "b", 1, func(rs []client.KeyValue) error {
> ```

Also run one with a page size of 2. You should be able to reuse most of the code (except this variable and the expected page size) and perhaps make two subtests.

---

*[pkg/internal/client/txn.go, line 506 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LGFBtj50ZrXjI0NRVgt:-LGFBtj50ZrXjI0NRVgu:byu28zj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67d7b4c612e2e037ee0fe29a2ccff709738eb22/pkg/internal/client/txn.go#L506)):*
> ```Go
> 			return errors.Wrap(err, "scanning meta2 keys")
> 		}
> 		if err := f(rows); err != nil {
> ```

Avoid calling `f` when there are no rows. This also avoids the awkward `+1` in your test.

---

*[pkg/server/admin.go, line 1303 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LGFC-Ro1HvhHUl7UtWX:-LGFC-Ro1HvhHUl7UtWY:b-l1sa9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67d7b4c612e2e037ee0fe29a2ccff709738eb22/pkg/server/admin.go#L1303)):*
> ```Go
> 	}
> 	if err := s.server.db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
> 		return txn.Iterate(ctx, keys.Meta2Prefix, keys.MetaMax, 10000, func(rows []client.KeyValue) error {
> ```

nit: factor out the `10000` into a `const pageSize` (can be defined just in the line above)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-:-LGFC8zH3y1EiqMGCqiC:bvsyumd)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: :shipit: complete! 1 of 0 LGTMs obtained

---

*[pkg/server/admin.go, line 1302 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LGFGVOe80H5QyzDNtPm:-LGFGVOf9cgbEvYMnaeg:bno7p14) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67d7b4c612e2e037ee0fe29a2ccff709738eb22/pkg/server/admin.go#L1302)):*
> ```Go
> 		replicaCounts[nodeID] = 0
> 	}
> 	if err := s.server.db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
> ```

Note that `Txn()` auto-retries the transaction if necessary. Its doc says this and that the function should not perform side effects.

This means the in-place count increase below is erroneous:

- some ranges will be counted mulitple times
- if the first txn attempt finds a range, then the range does not exist by the time of the 2nd txn attempt, then you'lll have a non-zero count remaining in `replicaCounts` for a range that effectively has zero replicas remaining.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 4 files at r4.
Review status: :shipit: complete! 1 of 0 LGTMs obtained

---

*[pkg/internal/client/txn.go, line 497 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LGJzj3S4DtGRlHDfYcO:-LGJzj3S4DtGRlHDfYcP:b-abj9jh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8809a5aa4653864ebda53955ad3c35993a54a064/pkg/internal/client/txn.go#L497)):*
> ```Go
> 
> // Iterate performs a paginated scan and applying the function f to every page.
> // The semantics of retrieval and ordering are the same as for Scan.
> ```

Mention restarts here in the comment - the paginated data must not be used for side effects before the txn has committed.

---

*[pkg/server/admin.go, line 1302 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LGFGVOe80H5QyzDNtPm:-LGJz7InBnZ9bSLlrkqp:bgk354v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67d7b4c612e2e037ee0fe29a2ccff709738eb22/pkg/server/admin.go#L1302)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Note that `Txn()` auto-retries the transaction if necessary. Its doc says this and that the function should not perform side effects.

This means the in-place count increase below is erroneous:

- some ranges will be counted mulitple times
- if the first txn attempt finds a range, then the range does not exist by the time of the 2nd txn attempt, then you'lll have a non-zero count remaining in `replicaCounts` for a range that effectively has zero replicas remaining.

</blockquote></details>

That's a very good point, Raphael. The code should be

```go
var replicaCounts map[roachpb.NodeID]int64

if err := Txn(..., func(txn *client.Txn) error {
  replicaCounts = make(map[roachpb.NodeID]int64)
  return txn.Iterate(...)
})

```

I saw that you tried to address this via `fInit`, but I'm not convinced this is the better API.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027)*
<!-- Sent from Reviewable.io -->

neeral commented:
Thanks for the feedback @tschottdorf and @knz. I've updated this commit.

---

Review status: :shipit: complete! 1 of 0 LGTMs obtained

---

*[pkg/cli/node.go, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LG6SL6-EDh0D-DqEgTs:-LGIrdi01Er1AszIpAgf:b5fxlia) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e654157fc963ca6e05b5c5a0934ca5438982840/pkg/cli/node.go#L320)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

use `stderr` defined in this package (not `os.Stderr`), this is necessary to test the output properly.
</blockquote></details>

I've changed to `stderr`. The tests require tweaking to verify this, though. They currently only check stdout.

---

*[pkg/internal/client/db_test.go, line 280 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LGFBKV3AnRbrThsvclb:-LGIvMp9A_jFtuVtWaNr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67d7b4c612e2e037ee0fe29a2ccff709738eb22/pkg/internal/client/db_test.go#L280)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Also run one with a page size of 2. You should be able to reuse most of the code (except this variable and the expected page size) and perhaps make two subtests.
</blockquote></details>

Done.

---

*[pkg/internal/client/txn.go, line 506 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LGFBtj50ZrXjI0NRVgt:-LGIvOelEU-G0EdorjgX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67d7b4c612e2e037ee0fe29a2ccff709738eb22/pkg/internal/client/txn.go#L506)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Avoid calling `f` when there are no rows. This also avoids the awkward `+1` in your test.
</blockquote></details>

Done.

---

*[pkg/internal/client/txn.go, line 497 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LGJzj3S4DtGRlHDfYcO:-LGMlKrx27uHW7-OtQss:b-sf1lo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8809a5aa4653864ebda53955ad3c35993a54a064/pkg/internal/client/txn.go#L497)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Mention restarts here in the comment - the paginated data must not be used for side effects before the txn has committed.
</blockquote></details>

I've updated the method comment.

---

*[pkg/server/admin.go, line 1302 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LGFGVOe80H5QyzDNtPm:-LGJ04oH3tXAS2hSy05r:b-xldpsu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67d7b4c612e2e037ee0fe29a2ccff709738eb22/pkg/server/admin.go#L1302)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

That's a very good point, Raphael. The code should be

```go
var replicaCounts map[roachpb.NodeID]int64

if err := Txn(..., func(txn *client.Txn) error {
  replicaCounts = make(map[roachpb.NodeID]int64)
  return txn.Iterate(...)
})

```

I saw that you tried to address this via `fInit`, but I'm not convinced this is the better API.
</blockquote></details>

I hadn't considered this. Thanks for pointing this out. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-:-LGMlrg48BvwkdYMiUWI:bnl1ag)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r5.
Review status: :shipit: complete! 1 of 0 LGTMs obtained

---

*[pkg/internal/client/txn.go, line 498 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LGOda7eEvqR_KBB3Wwc:-LGOda7fA6KmU_tg88Bi:b-ljcyxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f416d1b1973dd069b0045e0727372c85688e4a/pkg/internal/client/txn.go#L498)):*
> ```Go
> // Iterate performs a paginated scan and applying the function f to every page.
> // The semantics of retrieval and ordering are the same as for Scan. Note that
> // Txn auto-retries if the transaction if necessary. Hence, the paginated data
> ```

s/if//

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-:-LGOdfq-BxgN5Gx6saCx:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo the typo in the comment. Thank you for this contribution!

---

Reviewed 1 of 5 files at r2, 3 of 7 files at r3, 4 of 4 files at r5.
Review status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-:-LGPQUIY2t8SjmTGn_O2:b4t4kie)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
The typo is ` This ensures CockroachDB checks ranges <are>`, let me know when that's fixed and I'll merge!

neeral commented:
Review status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/internal/client/txn.go, line 498 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-LGOda7eEvqR_KBB3Wwc:-LGQqhQV4LTPPmGkNO4Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f416d1b1973dd069b0045e0727372c85688e4a/pkg/internal/client/txn.go#L498)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

s/if//
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
bors r+

Thanks @neeral!

---

Reviewed 6 of 6 files at r6.
Review status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27027#-:-LGTUH315LpO4Y1i00yS:b6evn8j)*
<!-- Sent from Reviewable.io -->
